feat: Add basic DW tables configuration for revenue

rafaeelaudibert opened the pull request:
We're now letting customers select a specific data warehouse table to display as revenue. This is not very useful yet because Stripe is the only integration that actually makes sense as revenue, but it'll grow eventually.

We're not using that information in Web Analytics yet, it's coming soon. There's a small problem where we'll need to handle rows/events that don't contain a session ID, so that might be slightly tricky

https://www.loom.com/share/ceac4d6b192e45f8a5b935e1d9a05219?sid=094dcc62-f9f7-48ab-b948-ed162dc8f77f


## Does this work well for both Cloud and self-hosted?

If you set up data warehouse, yeah

## How did you test this code?

Manually + unit + hoping to test it end-to-end in prod

rafaeelaudibert committed:
feat: Add basic data exchange tables configuration

We're now letting customers select a specific data warehouse table to display as revenue. This is not very useful yet because Stripe is the only integration that actually makes sense as revenue, but it'll grow eventually.

We're not using that information in the backend yet either, it's coming soon. There's a small problem where we'll need to handle rows/events that don't contain a session ID, so that might be slightly tricky.

rafaeelaudibert committed:
feat: Display revenue rows from our database

We're now querying the DB and looking for events in the DW tables. Main
concern right now is how do we limit it/sort by a specific timestamp to
get this to work/paginate properly? We might need to find a way to
figure out what the right "cursor" for a DW table is

rafaeelaudibert commented on frontend/src/queries/schema/schema-general.ts:
```diff
@@ -2688,7 +2704,28 @@ export enum CurrencyCode {
+}
+
 export interface RevenueTrackingConfig {
     baseCurrency?: CurrencyCode
-    events: RevenueTrackingEventItem[]
+
+    /*
+     * @default []
+     */
+    events?: RevenueTrackingEventItem[]
+
+    /*
+     * @default []
+     */
+    externalDataSchemas?: RevenueTrackingExternalDataSchema[]
 }
```
@robbie-c I'm actually curious here. How did you make `events` a required field if the underlying JSON is empty most of the time?

robbie-c replied:
The revenueTrackingConfig property on Teams was optional, so making its fields optional wasn't necessary then (whereas now it will be)

greptile's comment is just wrong, I think your approach here is totally correct.

rafaeelaudibert replied:
The catch is that I'd rather have them be not-nullable - of course. I'll see if pydantic can help us here :)

robbie-c replied:
If you're adding a new property it will probably always need to be nullable, at least on the TS side =/

rafaeelaudibert replied:
See with your own eyes :)

rafaeelaudibert committed:
fix: Fix `isEmpty` check

rafaeelaudibert committed:
feat: Explain new `ast.SelectQuery.empty()` class method

rafaeelaudibert committed:
fix: Improve query type

rafaeelaudibert committed:
fix: Handle `SelectSetQuery` properly on paginator

That wasn't handled properly before, pagination wasn't being applied as we expected because it didn't know how to handle a `SelectSetQuery`. Of course, pagination is still a little bit weird with UNION queries, but the current behavior is ok-ish

rafaeelaudibert committed:
fix: Avoid iterating on None

Events might be empty

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

rafaeelaudibert commented:
Tom showed me how they're doing this on Product Analytics and it's a much better solution to what I have here, I'll stop work on this and use their approach instead. I'll still use the same PR, but will move this to draft for now

rafaeelaudibert committed:
fix: Use much improved Data Warehouse selector

This looks much nicer now, it's based on how this works for Product Analytics, and the behavior is definitely much nicer. Backend was kept basically the same, only change is that we altered the property from `name` to `tableName` to be both more descriptive AND match the existing `eventName` for events

We've changed how pydantic/the Team model represent our schema slightly, the behavior is much nicer now!

rafaeelaudibert commented:
@lricoy @robbie-c Ready for review, might still have some tests breaking, will fix them tomorrow morning

https://www.loom.com/share/ceac4d6b192e45f8a5b935e1d9a05219?sid=094dcc62-f9f7-48ab-b948-ed162dc8f77f

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

rafaeelaudibert committed:
refactor: Improve type checking

rafaeelaudibert committed:
fix: Invert check order

rafaeelaudibert committed:
refactor: Avoid confusion by noticing it's updated inplace

rafaeelaudibert committed:
fix: Fix Python types

rafaeelaudibert committed:
fix: Fix TS types

rafaeelaudibert committed:
tests: Fix FE and BE tests (snapshots)

rafaeelaudibert committed:
fix: Properly cast to `Decimal64`

A couple months ago we changed our `toDecimal` function to use `toDecimal64OrNull` but that brought us a problem where it only accepts strings as first argument, we can't do something as simple as `toDecimal(10, 1)` because it'd complain the first argument is a Float.

To fix that, let's go back to using `accurateCastOrNull`. The downside is that the way we decide on the scale is pretty weird, it looks like this `accurateCastOrNull(10, 'Decimal64(1)')` inside a string. We can easily add the extra string parameter by using `suffixArgs` (which only really exists because of `accurateCastOrNull` usage) but we couldn't easily add a value inside of it. That's solved now. It's full of restrictions (you can only pass constants, for example) but it works pretty well, see regenerated snapshots.

While I was at that, I changed `DATE` to `_toDate` to match our other functions, this was a mistake from the past.

rafaeelaudibert committed:
test: Skip tests in CI

They break pretty bad with a CH error

rafaeelaudibert committed:
fix: Fix python type error

rafaeelaudibert committed:
feat: Introduce required `revenueCurrencyProperty`

It's hard to test this feature out if you can't say, for example, that
all of the events coming from a specific type of event have a set
currency that's different than the base currency. We could allow people
to write HogQL here, but this is a first step towards that.

I've updated both UI, server and tests in a single commit. I know this
shouldn't fit on the Data Warehouse PR, but it's the easiest way to get
this out and not have tons and tons of PRs open at the same time.

rafaeelaudibert committed:
fix: Disable stories

They keep flapping not sure why, I let the Product Analytics team know

rafaeelaudibert committed:
tests: Fix snapshot

rafaeelaudibert commented on frontend/src/lib/api.mock.ts:
```diff
@@ -99,13 +100,15 @@ export const MOCK_DEFAULT_TEAM: TeamType = {
                 eventName: 'purchase',
                 revenueProperty: 'value',
-                revenueCurrencyProperty: undefined,
+                revenueCurrencyProperty: { static: CurrencyCode.ZAR },
```
Used ZAR because South Africa's flag looks cool on the snapshot and you must be tired of me adding the Brazilian flag everywhere :)

lricoy replied:
Are we though? 🇧🇷

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

rafaeelaudibert committed:
fix: Last tests fixes plz

rafaeelaudibert commented:
@robbie-c @lricoy here's a snippet you can run your local Django shell to test this

```python
team = Team.objects.last()
source = ExternalDataSource.objects.create(
    team=team,
    source_id="source_id_1",
    connection_id="connection_id_1",
    status=ExternalDataSource.Status.COMPLETED,
    source_type=ExternalDataSource.Type.STRIPE,
)
credentials = DataWarehouseCredential.objects.create(access_key="blah", access_secret="blah", team=team)
warehouse_table = DataWarehouseTable.objects.create(
    name="table_1",
    format="Parquet",
    team=team,
    external_data_source=source,
    external_data_source_id=source.id,
    credential=credentials,
    url_pattern="https://bucket.s3/data/*",
    columns={
        'id': {'hogql': 'StringDatabaseField', 'clickhouse': 'Nullable(String)', 'schema_valid': True},
        'revenue': {'hogql': 'StringDatabaseField', 'clickhouse': 'Float64', 'schema_valid': True},
        'timestamp': {'hogql': 'DateTimeDatabaseField', 'clickhouse': 'Nullable(DateTime)', 'schema_valid': True},
    },
)
ExternalDataSchema.objects.create(
    team=team,
    name="table_1",
    source=source,
    table=warehouse_table,
    should_sync=True,
    last_synced_at="2024-01-01",
)
```

rafaeelaudibert committed:
refactor: Rename to data warehouse tables on revenue

We were calling these external tables when talking about revenue, let's call them "data warehouse tables" instead, it's much more reasonable for customers

rafaeelaudibert committed:
feat: Support static/dynamic currency for data warehouse tables

Recently added this for events, let's add it for tables too

rafaeelaudibert committed:
fix: Revert incorrect changes to api

rafaeelaudibert committed:
fix: Fix snapshot again aa

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

rafaeelaudibert commented on frontend/__snapshots__/scenes-app-data-management--revenue-events-settings--dark.png:
Hmm, this looks much nicer on a bigger screen, will try and adjust it slightly

lricoy reviewed (approved):
Nit suggestions for future PRs:

`ControlledDefinitionPopover`
- Make the required fields stand out more (maybe make them bold)
- Sort the fields by type or maybe just accept timestamp type fields on the `timestamp` option? I am not sure if we could make timestamp strings work.

lricoy reviewed (approved):
Nit suggestions for future PRs:

`ControlledDefinitionPopover`
- Make the required fields on the stand out more (maybe make them bold)
- Sort the fields by type or maybe just accept timestamp type fields on the `timestamp` option? I am not sure if we could make timestamp strings work.

lricoy reviewed (approved):
Nit suggestions for future PRs:

`ControlledDefinitionPopover`
- Make the required fields on the stand out more (maybe make them bold)
- Sort the fields by type or maybe just accept timestamp type fields on the `timestamp` option? I am not sure if we could make timestamp strings work.

lricoy reviewed (approved):
Nit suggestions for future PRs:

`ControlledDefinitionPopover`
- Make the required fields on the stand out more (maybe make them bold)
- Sort the fields by type or maybe just accept timestamp type fields on the `timestamp` option? I am not sure if we could make timestamp strings work.
